File: boat-sample/src/test/java/sample/java/xyz/srclab/common/bus/EventBusSample.java
package sample.java.xyz.srclab.common.bus;

import org.jetbrains.annotations.NotNull;
import org.testng.annotations.Test;
import xyz.srclab.common.bus.EventBus;
import xyz.srclab.common.bus.EventHandler;
import xyz.srclab.common.bus.EventHandlerNotFoundException;
import xyz.srclab.common.test.TestLogger;

import java.util.Arrays;

public class EventBusSample {

    private static final TestLogger logger = TestLogger.DEFAULT;

    @Test
    public void testEventBus() {
        EventBus eventBus = EventBus.newEventBus(Arrays.asList(
            new EventHandler<Object>() {
                @NotNull
                @Override
                public Object eventType() {
                    return String.class;
                }

                @Override
                public void handle(@NotNull Object event) {
                    logger.log(event);
                }
            },
            new EventHandler<Object>() {
                @NotNull
                @Override
                public Object eventType() {
                    return Integer.class;
                }

                @Override
                public void handle(@NotNull Object event) {
                    logger.log(event);
                }
            }
        ));
        //1
        eventBus.emit(1);
        //2
        eventBus.emit("2");
        //No output
        eventBus.emit(new Object());
        try {
            eventBus.emitOrThrow(new Object());
        } catch (EventHandlerNotFoundException e) {
            //xyz.srclab.common.bus.EventHandlerNotFoundException: class java.lang.Object
            logger.log(e);
        }
    }
}

File: boat-sample/src/test/kotlin/sample/kotlin/xyz/srclab/common/bus/EventBusSample.kt
package sample.kotlin.xyz.srclab.common.bus

import org.testng.annotations.Test
import xyz.srclab.common.bus.EventBus
import xyz.srclab.common.bus.EventHandler
import xyz.srclab.common.bus.EventHandlerNotFoundException
import xyz.srclab.common.test.TestLogger

class EventBusSample {

    @Test
    fun testEventBus() {
        val eventBus = EventBus.newEventBus(
            listOf(
                object : EventHandler<Any> {

                    override val eventType: Any
                        get() {
                            return String::class.java
                        }

                    override fun handle(event: Any) {
                        logger.log(event)
                    }
                },
                object : EventHandler<Any> {

                    override val eventType: Any
                        get() {
                            return Int::class.java
                        }

                    override fun handle(event: Any) {
                        logger.log(event)
                    }
                }
            ))
        //1
        eventBus.emit(1)
        //2
        eventBus.emit("2")
        //No output
        eventBus.emit(Any())
        try {
            eventBus.emitOrThrow(Any())
        } catch (e: EventHandlerNotFoundException) {
            //xyz.srclab.common.bus.EventHandlerNotFoundException: class java.lang.Object
            logger.log(e)
        }
    }

    companion object {
        private val logger = TestLogger.DEFAULT
    }
}